pkg/server/plugin/bundlepublisher/gcpcloudstorage/gcpcloudstorage_test.go

```diff
@@ -0,0 +1,427 @@
+				ObjectName: "object-name",
+				Format:     "spiffe",
+			},
+		},
+		{
+			name:   "write failure",
+			bundle: testBundle,
+			config: &Config{
+				BucketName: "bucket-name",
+				ObjectName: "object-name",
+				Format:     "spiffe",
+			},
+			writeErr:   errors.New("write error"),
+			expectCode: codes.Internal,
+			expectMsg:  "failed to write bundle: write error",
+		},
+		{
+			name:   "close failure",
+			bundle: testBundle,
+			config: &Config{
+				BucketName: "bucket-name",
+				ObjectName: "object-name",
+				Format:     "spiffe",
+			},
+			closeErr:   errors.New("close error"),
+			expectCode: codes.Internal,
+			expectMsg:  "failed to close storage writer: close error",
+		},
+		{
+			name:       "not configured",
+			expectCode: codes.FailedPrecondition,
+			expectMsg:  "not configured",
+		},
+		{
+			name: "missing bundle",
+			config: &Config{
+				BucketName: "bucket-name",
+				ObjectName: "object-name",
+				Format:     "spiffe",
+			},
+			expectCode: codes.InvalidArgument,
+			expectMsg:  "missing bundle in request",
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			var err error
+			options := []plugintest.Option{
+				plugintest.CaptureConfigureError(&err),
+				plugintest.CoreConfig(catalog.CoreConfig{
+					TrustDomain: spiffeid.RequireTrustDomainFromString("example.org"),
+				}),
+				plugintest.ConfigureJSON(tt.config),
+			}
+
+			newClient := func(ctx context.Context, opts ...option.ClientOption) (gcsService, error) {
+				return &fakeClient{
+					clientOptions: opts,
+				}, nil
+			}
+
+			newStorageWriter := func(ctx context.Context, o *storage.ObjectHandle) io.WriteCloser {
+				return &fakeStorageWriter{
+					writeErr: tt.writeErr,
+					closeErr: tt.closeErr,
+				}
+			}
+			p := newPlugin(newClient, newStorageWriter)
+
+			if tt.config != nil {
+				plugintest.Load(t, builtin(p), nil, options...)
+				require.NoError(t, err)
+			}
+
+			resp, err := p.PublishBundle(context.Background(), &bundlepublisherv1.PublishBundleRequest{
+				Bundle: tt.bundle,
+			})
+
+			if tt.expectMsg != "" {
+				spiretest.RequireGRPCStatusContains(t, err, tt.expectCode, tt.expectMsg)
+				return
+			}
+			require.NoError(t, err)
+			require.NotNil(t, resp)
+		})
+	}
+}
+
+func TestPublishMultiple(t *testing.T) {
+	config := &Config{
+		BucketName: "bucket-name",
+		ObjectName: "object-name",
+		Format:     "spiffe",
+	}
+
+	var err error
+	options := []plugintest.Option{
+		plugintest.CaptureConfigureError(&err),
+		plugintest.CoreConfig(catalog.CoreConfig{
+			TrustDomain: spiffeid.RequireTrustDomainFromString("example.org"),
+		}),
+		plugintest.ConfigureJSON(config),
+	}
+
+	newClient := func(ctx context.Context, opts ...option.ClientOption) (gcsService, error) {
+		return &fakeClient{
+			clientOptions: opts,
+		}, nil
+	}
+	newStorageWriter := getFakeNewStorageWriterFunc(nil, nil)
+	p := newPlugin(newClient, newStorageWriter)
+	p.hooks.wroteObjectFunc = func() { writeObjectCount++ }
```

Comment:
it looks like this var `wroteObjectCount` is only used in this function, may we create it here?